lib/puppet/util/autoload.rb

```diff
@@ -163,7 +162,9 @@ def gem_directories
     end
 
     def search_directories(env)
-      [gem_directories, module_directories(env), libdirs(), $LOAD_PATH].flatten
+      # Ensure that module_directories have been added to $LOAD_PATH
+      module_directories(env)
```

Comment:
AFAICT, `search_directories` get recalculated every time. The `module_directories` however, is cached with the environment.